src/Player.cpp

```diff
@@ -10,90 +11,209 @@ Player::Player( sf::View& view, Map& map )
+	if ( hit_left_border() ) {
+		stop_moving();
 	}
 	// moving left, dont move view
-	else if ( m_animations.position().x > Config::Background::width - Config::Window::width / 2
-			  || m_animations.position().x < Config::Window::width / 2 ) {
+	else if ( near_left_border() || near_right_border() ) {
 		m_velocity.x -= Config::Player::acceleration.x;
-	} 
+	}
 	// move view also
 	else {
-		m_view.move( { -Config::View::acceleration, 0 } );
+		m_view_speed.x = -Config::View::acceleration;
 		m_velocity.x -= Config::Player::acceleration.x;
 	}
-	// change animation
-	m_current = m_animations.left_animation();
 }
-
 void Player::move_up() {
+
 	// falling down
-	if ( m_acceleration <= 0 ) { 
+	if ( m_acceleration <= 0 ) {
 		m_is_falling = true; 
-		std::cout << m_animations.position().y << std::endl;
 	}
-	// hit ground
-	if ( m_animations.position().y == 416 && m_is_falling ) {
+
+	const auto& map = m_map.map();
+	auto new_pos_y = position().y + m_velocity.y;
+	auto index_x_left = (int)( player_left_border(position()) ) / Config::Block::size;
+	auto index_x_right = (int)( player_right_border(position()) ) / Config::Block::size;
+	auto index_y = (int)( Config::Window::height - new_pos_y - 25 ) / 32;
+
+	// ground or block under
+	if ( m_is_falling 
+		 && ( index_y == 0 
+		 ||  map[ index_y - 1 ][ index_x_left ] != nullptr 
+		 ||  map[ index_y - 1 ][ index_x_right ] != nullptr )) {
+		// hit the ground
+		if ( index_y == 0 ) {
+			m_animations.animated_sprite().setPosition( position().x, 416 );
+		}
+		// hit the block
+		else {
+			sf::Vector2f pos;
+			//get the position of the block
+			if ( m_map.block_exists_at( index_y - 1, index_x_left )) {
+				pos = map[ index_y - 1 ][ index_x_left ]->position();
+			}
+			if ( m_map.block_exists_at( index_y - 1, index_x_right ) ) {
+				pos = map[ index_y - 1 ][ index_x_right ]->position();
+			}
+			// place player above the block
+			m_animations.animated_sprite().setPosition( position().x, pos.y - 64 );
+		}
+
+
 		m_is_jumping = false;
 		m_is_falling = false;
+		m_velocity.y = 0;
 		m_acceleration = Config::Player::acceleration.y;
 	}
 	// going up
 	if ( m_is_jumping ) {
 		m_velocity.y -= m_acceleration;
 		m_acceleration -= Config::Physics::gravity;
 	}
+	
 }
 
-void Player::move() {
+void Player::fall_down() {
+	// falling
+	if ( m_is_falling_from_block ) {
+		m_velocity.y += m_acceleration2;
+		m_acceleration2 += Config::Physics::gravity;
+	}
+	const auto& map = m_map.map();
+	auto new_pos_y = position().y + m_velocity.y;
+
+	// index depends whether we go right or left
+	int index_x;
+	if ( m_direction == Directions::Right ) {
+		index_x = (int)(player_right_border(position()) - 32) / 32;
+	}
+	if ( m_direction == Directions::Left ) {
+		index_x = (int)(player_left_border(position()) - 32) / 32;
+	}
+	auto index_y = (int)(Config::Window::height - new_pos_y - 25) / 32;
+	// ground or block under
+	if ( m_is_falling_from_block && (index_y == 0 || map[ index_y - 1 ][ index_x ] != nullptr) ) {

```

Comment:
Here you can take out  the part of condition`(index_y == 0 || map[ index_y - 1 ][ index_x ] != nullptr` and insert it into a nested if-else.